### PR TITLE
Add support for crashtests

### PR DESCRIPTION
**Pull request by jgraham**
See https://github.com/web-platform-tests/rfcs/pull/33

### CODE REVIEW

**Comment by jgraham** [comment]
@emilio you might also be interested in this.

**Review by marcoscaceres — APPROVED** [review]
approving payment request test rename. Thanks!

**Review by FremyCompany — APPROVED** [review]
At first glance, the css-tables related renames seem ok

**Review by svgeesus — APPROVED** [review]
The renames to avoid -crash where it shouldn't be seem good. I skimmed over the python changes but someone else should review those.

File: tools/manifest/sourcefile.py [hunk 1]
```diff
@@ -408,6 +408,11 @@ def name_is_reference(self):
         be a reference file (not a reftest)"""
         return "/reference/" in self.url or bool(reference_file_re.search(self.name))
 
+    @property
+    def name_is_crashtest(self):
+        # type: () -> bool
+        return self.type_flag == "crashtest" or "crashtests" in self.dir_path.split(os.path.sep)
```
Comment (gsnedders) [inline tools/manifest/sourcefile.py]: RFC as approved uses `crash` for the `type_flag`.

**Review by gsnedders — CHANGES_REQUESTED** [review]
Files that shouldn't have been renamed:

 * adjoining-float-new-fc-thcrash.html (doesn't meet the criteria in the commit message)
 * gradient-thcrash.html (is a reftest?!)
 * extremely-tall-multicol-with-extremely-tall-child-thcrash.html (doesn't meet the criteria in the commit message)
 * before-in-display-none-thcrash.html (ditto)
 * shape-outside-inset-thcrash.html (is a reftest)

I'm not gonna bother going through the rest, something has gone wrong…

**Review by gsnedders — CHANGES_REQUESTED** [review]
`css/css-position/position-absolute-crash-chrome-005.html` and `css/css-scoping/whitespace-crash-001.html` don't need renamed (they're not crash tests).

I also wonder if we should have a lint for `-crash` tests containing testharness.js? It really is an antipattern, but also something Chrome developers have already done (and AFAIK they run differently as a result in `run-webkit-tests`, @Hexcles?).

**Comment by jgraham** [comment]
> `css/css-position/position-absolute-crash-chrome-005.html` and `css/css-scoping/whitespace-crash-001.html` don't need renamed (they're not crash tests).

Fixed.

**Comment by Hexcles** [comment]
> css/css-position/position-absolute-crash-chrome-005.html and css/css-scoping/whitespace-crash-001.html don't need renamed (they're not crash tests).

They could be rewritten to real tests though, but that's for another day.

> I also wonder if we should have a lint for -crash tests containing testharness.js?

I agree. We should.

> It really is an antipattern, but also something Chrome developers have already done (and AFAIK they run differently as a result in run-webkit-tests, @Hexcles?).

Sorry I'm not sure what you're asking? Until today's introduction of crash tests, Chrome developers have written some de facto crash tests with dummy testharness asserts in them. These should not be needed in the future (at least not named with `-crash`). However, it is true that `run_web_tests` might need to be updated to support these assert-free crash tests (well, more importantly the lack of `testharnessreport.js`, due to a hack we use internally).

**Comment by jgraham** [comment]
>> I also wonder if we should have a lint for -crash tests containing testharness.js?
> I agree. We should.

I think we could leave that to a followup? It doesn't seem critical to landing the patch, and there are already a bunch of exceptions which we could/should fix at the same time, but deciding how to proceed with that is more work.

**Comment by gsnedders** [comment]
> Until today's introduction of crash tests, Chrome developers have written some de facto crash tests with dummy testharness asserts in them.

AIUI, does run-webkit-tests not run `-crash` tests in WPT as actual crash tests already? i.e., it ignores any asserts (or failing asserts) in them. My concern is people adding new tests like that with actual asserts.

But to me I'm happy to leave it as a followup.

**Review by gsnedders — APPROVED** [review]

**Comment by Hexcles** [comment]
> AIUI, does run-webkit-tests not run -crash tests in WPT as actual crash tests already? i.e., it ignores any asserts (or failing asserts) in them. My concern is people adding new tests like that with actual asserts.

That's new to me, but it's possible. And yes we should prevent people from adding dummy asserts to crash tests.

**Pull request merged**